File: apps/admin/src/api/stats.js
import { api } from './http'
import { toInteger, toNumber } from '../utils/format'
import { CHARGER_STATUS } from '../utils/domain'

/**
 * 统计接口（Go 契约 GET /admin/stats/revenue 与 /admin/stats/chargers）。
 *
 * 旧 Qt 管理端的统计来自服务端预计算的驾驶舱快照，Go 侧改为按需聚合：
 * 营收按「已完成订单的冻结金额、归属停止充电时刻」统计，分桶对齐请求区间起点，
 * 空区间也会返回零值桶，图表因此不会把安静的日子跳过。
 *
 * 服务端返回整数分与整数毫瓦时；这里只做契约校验，不做任何隐式换算，
 * 单位换算全部留给展示层。
 */

/** 响应体结构不符合契约时抛出的错误，携带可直接展示的 userMessage。 */
export class DtoError extends Error {
  constructor(message) {
    super(message)
    this.name = 'DtoError'
    this.userMessage = message
  }
}

/**
 * 营收统计：fromAt/toAt 为 UTC 秒且必填，单次区间上限 90 天；
 * stationId 可选，bucket 为 day（趋势）或 hour（单日明细）。
 * 解析交给 parseRevenueStats，使 DTO 校验只有一处。
 */
export function fetchRevenueStats({ fromAt, toAt, stationId, bucket } = {}) {
  return api.get('/admin/stats/revenue', { fromAt, toAt, stationId, bucket })
}

/** 设备状态统计；stationId 可选。解析交给 parseChargerStatusStats。 */
export function fetchChargerStatusStats({ stationId } = {}) {
  return api.get('/admin/stats/chargers', { stationId })
}

function requireInteger(value, field) {
  const parsed = toInteger(value)
  if (parsed === null) throw new DtoError(`${field} 字段缺失或格式不正确`)
  return parsed
}

function requireNumber(value, field) {
  const parsed = toNumber(value)
  if (parsed === null) throw new DtoError(`${field} 字段缺失或格式不正确`)
  return parsed
}

/** 解析营收统计响应：items 与三个总计字段都必须存在且为整数。 */
export function parseRevenueStats(data) {
  if (!data || typeof data !== 'object') throw new DtoError('营收统计响应为空')
  if (!Array.isArray(data.items)) throw new DtoError('营收统计响应缺少 items 列表')
  const items = data.items.map((item, index) => {
    if (!item || typeof item !== 'object') throw new DtoError(`营收统计第 ${index + 1} 条记录格式不正确`)
    return {
      bucketStart: requireInteger(item.bucketStart, `items[${index}].bucketStart`),
      amountCent: requireInteger(item.amountCent, `items[${index}].amountCent`),
      energyMwh: requireInteger(item.energyMwh, `items[${index}].energyMwh`),
      orderCount: requireInteger(item.orderCount, `items[${index}].orderCount`)
    }
  })
  return {
    items,
    totalAmountCent: requireInteger(data.totalAmountCent, 'totalAmountCent'),
    totalEnergyMwh: requireInteger(data.totalEnergyMwh, 'totalEnergyMwh'),
    totalOrderCount: requireInteger(data.totalOrderCount, 'totalOrderCount')
  }
}

/** 解析设备状态统计响应：五种状态数量、可运营数、总数与健康度都必须存在。 */
export function parseChargerStatusStats(data) {
  if (!data || typeof data !== 'object') throw new DtoError('设备状态统计响应为空')
  const stats = {
    idleCount: requireInteger(data.idleCount, 'idleCount'),
    occupiedCount: requireInteger(data.occupiedCount, 'occupiedCount'),
    faultyCount: requireInteger(data.faultyCount, 'faultyCount'),
    restartingCount: requireInteger(data.restartingCount, 'restartingCount'),
    disabledCount: requireInteger(data.disabledCount, 'disabledCount'),
    operationalCount: requireInteger(data.operationalCount, 'operationalCount'),
    totalCount: requireInteger(data.totalCount, 'totalCount'),
    healthPercent: requireNumber(data.healthPercent, 'healthPercent')
  }
  if (stats.totalCount < 0 || stats.operationalCount < 0) throw new DtoError('设备状态统计数量不能为负数')
  return stats
}

/** 空统计：接口未返回数据时保持页面结构稳定，而非渲染 NaN。 */
export function emptyChargerStatusStats() {
  return {
    idleCount: 0,
    occupiedCount: 0,
    faultyCount: 0,
    restartingCount: 0,
    disabledCount: 0,
    operationalCount: 0,
    totalCount: 0,
    healthPercent: 0
  }
}

/** 把统计拆成图表用的分段（顺序固定为 空闲/在用/故障/已停用/重启中）。 */
export function chargerStatusBreakdown(stats) {
  if (!stats) return []
  const counts = {
    idle: stats.idleCount,
    occupied: stats.occupiedCount,
    faulty: stats.faultyCount,
    disabled: stats.disabledCount,
    restarting: stats.restartingCount
  }
  return CHARGER_STATUS.map(item => ({
    key: item.key,
    label: item.label,
    tone: item.tone,
    value: toInteger(counts[item.key]) ?? 0
  }))
}

File: backend/internal/admin/stats.go
package admin

import (
	"context"
	"errors"
	"fmt"
	"math"
)

// The operations statistics the console reads.
//
// These are the Go replacement for the analytics the old service precomputed
// into a dashboard snapshot. The snapshot is gone and the numbers are computed
// on demand, because a precomputed snapshot has to be rebuilt on a timer, goes
// stale between rebuilds, and answers with data that may be hours old - which
// is a poor trade for an aggregate PostgreSQL answers in milliseconds.
//
// Two definitions are carried over from the old service unchanged, because they
// are business rules rather than implementation details:
//
//   - Revenue is the bill of a COMPLETED order, attributed to the moment the
//     charge stopped. That is when the amount was frozen, and it is the only
//     timestamp that cannot change afterwards.
//   - A charger is operational when it is idle or occupied. A faulty or
//     disabled device cannot serve, and a restarting one is not available at
//     the moment it is counted.

// Bucket names for the revenue series.
const (
	StatsBucketHour = "hour"
	StatsBucketDay  = "day"
)

const (
	secondsPerHour = int64(3600)
	secondsPerDay  = int64(86400)

	// MaxStatsRangeSeconds bounds one statistics query at 90 days, matching the
	// window the console offers and the ML feature window. An unbounded range on
	// an hour-bucketed series would return years of points to a chart that can
	// draw a few hundred.
	MaxStatsRangeSeconds = 90 * secondsPerDay
)

// Statistics errors. Handlers map all of them to 400: they describe a query the
// caller wrote, not a failure of the platform.
var (
	// ErrInvalidStatsRange reports a missing, inverted or over-long time range.
	ErrInvalidStatsRange = errors.New("admin: invalid statistics time range")
	// ErrInvalidStatsBucket reports a bucket outside the contract enum.
	ErrInvalidStatsBucket = errors.New("admin: invalid statistics bucket")
	// ErrInvalidStatsStation reports a station filter below 1.
	ErrInvalidStatsStation = errors.New("admin: invalid statistics station filter")
)

// RevenueQuery is a validated revenue request.
type RevenueQuery struct {
	// FromAt and ToAt are UTC Unix seconds. The range is half-open: FromAt is
	// included, ToAt is not.
	FromAt int64
	ToAt   int64
	// StationID filters to one station; 0 means every station.
	StationID int64
	// Bucket is StatsBucketHour or StatsBucketDay.
	Bucket string
}

// BucketSeconds is the width of one bucket in seconds.
func (q RevenueQuery) BucketSeconds() int64 {
	if q.Bucket == StatsBucketHour {
		return secondsPerHour
	}
	return secondsPerDay
}

// Validate reports whether the query is one the platform will run.
func (q RevenueQuery) Validate() error {
	if q.FromAt < 0 || q.ToAt <= q.FromAt {
		return fmt.Errorf("%w: FromAt must be non-negative and ToAt must be after it", ErrInvalidStatsRange)
	}
	if q.ToAt-q.FromAt > MaxStatsRangeSeconds {
		return fmt.Errorf("%w: the range may not exceed %d days", ErrInvalidStatsRange, MaxStatsRangeSeconds/secondsPerDay)
	}
	if q.Bucket != StatsBucketHour && q.Bucket != StatsBucketDay {
		return fmt.Errorf("%w: %q is not %s or %s", ErrInvalidStatsBucket, q.Bucket, StatsBucketHour, StatsBucketDay)
	}
	if q.StationID < 0 {
		return fmt.Errorf("%w: StationID must not be negative", ErrInvalidStatsStation)
	}
	return nil
}

// RevenueBucket is one point of the revenue series.
type RevenueBucket struct {
	// BucketStart is the UTC Unix second the bucket begins at.
	BucketStart int64
	AmountCent  int64
	EnergyMwh   int64
	OrderCount  int64
}

// RevenueTotals is the whole range, which is not always the sum of the buckets
// a caller receives: the buckets are capped, the totals are not.
type RevenueTotals struct {
	AmountCent int64
	EnergyMwh  int64
	OrderCount int64
}

// RevenueStats is the revenue series with its range totals.
type RevenueStats struct {
	Items           []RevenueBucket
	TotalAmountCent int64
	TotalEnergyMwh  int64
	TotalOrderCount int64
}

// ChargerCounts is the raw fleet census, as the store reports it.
type ChargerCounts struct {
	Idle       int64
	Occupied   int64
	Faulty     int64
	Restarting int64
	Disabled   int64
	Total      int64
}

// ChargerStatusStats is the fleet census with the derived health figures.
type ChargerStatusStats struct {
	IdleCount        int64
	OccupiedCount    int64
	FaultyCount      int64
	RestartingCount  int64
	DisabledCount    int64
	OperationalCount int64
	TotalCount       int64
	// HealthPercent is OperationalCount as a percentage of TotalCount, rounded
	// to two decimals; 0 when the fleet is empty.
	HealthPercent float64
}

// FleetCounts is the lifetime and current-state census behind the overview.
type FleetCounts struct {
	// TotalRevenueCent, CompletedOrders, FastOrders and SlowOrders are lifetime
	// figures over every completed order, not over a window.
	TotalRevenueCent int64
	CompletedOrders  int64
	FastOrders       int64
	SlowOrders       int64
	// RegisteredUsers counts every account the platform holds.
	RegisteredUsers int64
	// Stations counts every station, including a disabled one: an operator
	// asking how many stations exist is not asking how many are open.
	Stations int64
	// ActiveOrders counts the orders that have not reached a terminal state.
	ActiveOrders int64
}

// StatsOverview is the operations snapshot behind GET /admin/stats/overview.
//
// It is the scalar block the old dashboard snapshot carried: what the platform
// has, not what happened in a window. Windowed figures belong to
// GET /admin/stats/revenue, because only the client knows which timezone its
// "today" is in.
type StatsOverview struct {
	GeneratedAt         int64
	TotalRevenueCent    int64
	TotalChargeCount    int64
	FastChargeCount     int64
	SlowChargeCount     int64
	RegisteredUserCount int64
	StationCount        int64
	ActiveOrderCount    int64
	Chargers            ChargerStatusStats
}

// RevenueStats returns the revenue series and the range totals.
//
// The series is dense: every bucket in the range is present, with zeros where
// nothing happened. A chart that skipped the quiet days would draw a misleading
// trend, and a table that omitted them would look like missing data.
func (s *Service) RevenueStats(ctx context.Context, query RevenueQuery) (RevenueStats, error) {
	if err := query.Validate(); err != nil {
		return RevenueStats{}, err
	}
	rows, err := s.store.RevenueBuckets(ctx, query)
	if err != nil {
		return RevenueStats{}, err
	}
	totals, err := s.store.RevenueTotals(ctx, query)
	if err != nil {
		return RevenueStats{}, err
	}
	return RevenueStats{
		Items:           fillRevenueBuckets(query, rows),
		TotalAmountCent: totals.AmountCent,
		TotalEnergyMwh:  totals.EnergyMwh,
		TotalOrderCount: totals.OrderCount,
	}, nil
}

// ChargerStatus returns the fleet census with its derived health figures.
func (s *Service) ChargerStatus(ctx context.Context, stationID int64) (ChargerStatusStats, error) {
	if stationID < 0 {
		return ChargerStatusStats{}, fmt.Errorf("%w: station id must not be negative", ErrInvalidStatsStation)
	}
	counts, err := s.store.ChargerCounts(ctx, stationID)
	if err != nil {
		return ChargerStatusStats{}, err
	}
	return newChargerStatusStats(counts), nil
}

// Overview returns the operations snapshot.
func (s *Service) Overview(ctx context.Context) (StatsOverview, error) {
	counts, err := s.store.FleetCounts(ctx)
	if err != nil {
		return StatsOverview{}, err
	}
	chargers, err := s.store.ChargerCounts(ctx, 0)
	if err != nil {
		return StatsOverview{}, err
	}
	return StatsOverview{
		GeneratedAt:         s.clock().Unix(),
		TotalRevenueCent:    counts.TotalRevenueCent,
		TotalChargeCount:    counts.CompletedOrders,
		FastChargeCount:     counts.FastOrders,
		SlowChargeCount:     counts.SlowOrders,
		RegisteredUserCount: counts.RegisteredUsers,
		StationCount:        counts.Stations,
		ActiveOrderCount:    counts.ActiveOrders,
		Chargers:            newChargerStatusStats(chargers),
	}, nil
}

// newChargerStatusStats derives the operational count and the health figure.
//
// The rule is the old service's: idle and occupied devices are operational, the
// rest are not. The percentage keeps two decimals where the old one truncated
// to a whole number - the console shows a fraction of a percent, and throwing
// it away made a fleet that lost two devices look unchanged.
func newChargerStatusStats(counts ChargerCounts) ChargerStatusStats {
	operational := counts.Idle + counts.Occupied
	health := 0.0
	if counts.Total > 0 {
		health = math.Round(float64(operational)*10000/float64(counts.Total)) / 100
	}
	return ChargerStatusStats{
		IdleCount:        counts.Idle,
		OccupiedCount:    counts.Occupied,
		FaultyCount:      counts.Faulty,
		RestartingCount:  counts.Restarting,
		DisabledCount:    counts.Disabled,
		OperationalCount: operational,
		TotalCount:       counts.Total,
		HealthPercent:    health,
	}
}

// fillRevenueBuckets lays the stored buckets onto a dense series.
//
// Buckets are aligned to FromAt rather than to the epoch. The client computes
// its own day boundaries - its "today" starts at its local midnight - and asks
// for exactly that range, so aligning to the range is what makes each bucket
// one of the client's days. Epoch alignment would put the first bucket's
// boundary in the middle of a local day and show a partial day as a quiet one.
func fillRevenueBuckets(query RevenueQuery, rows []RevenueBucket) []RevenueBucket {
	width := query.BucketSeconds()
	stored := make(map[int64]RevenueBucket, len(rows))
	for _, row := range rows {
		stored[row.BucketStart] = row
	}

	items := make([]RevenueBucket, 0, (query.ToAt-query.FromAt)/width+1)
	for start := query.FromAt; start < query.ToAt; start += width {
		bucket, found := stored[start]
		if !found {
			bucket = RevenueBucket{}
		}
		bucket.BucketStart = start
		items = append(items, bucket)
	}
	return items
}
